File: SWEA/D3/2806. N－Queen/N－Queen.py
def check(row):
    for i in range(row):
        if board[i] == board[row] or abs(board[i] - board[row]) == abs(i-row):
            return False
    return True

def DFS(row):
    global answer
    if row == n:
        answer += 1
        return
    for col in range(n):
        board[row] = col
        if check(row):
            DFS(row+1)

t = int(input())
for i in range(1, t+1):
    n = int(input())
    board = [0] * n
    answer = 0

    DFS(0)

    print("#%d %d" % (i, answer))

File: 프로그래머스/2/12952. N－Queen/N－Queen.java
import java.util.*;

class Solution {
    private int[] board;
    private int n;
    private int answer = 0;
    
    public int solution(int n) {
        
        this.n = n;
        this.board = new int[n];
        
        dfs(0);
        
        return answer;
    }
    
    private void dfs(int row) {
        if (row == n) {
            answer++;
            return;
        }
        
        for (int i = 0; i < n; i++) {
            boolean isPossible = true;
            
            for (int j = 0; j < row; j++) {
                if (board[j] == i || Math.abs(row - j) == Math.abs(i - board[j])) {
                    isPossible = false;
                    break;
                }
            }
            
            if (isPossible) {
                board[row] = i;
                dfs(row + 1);
            }
        }
    }
}